### PR TITLE
Add product puppeteer class and simplify component update behavior

### PR DESCRIPTION
**Pull request by lhoffbeck**
⚠️  This PR builds on the work to support combined listings and 2k variants https://github.com/Shopify/dawn/pull/3378

tl;dr - this PR primarily copies code out of the variant picker class in `global.js` into `product-info.js`, which has been promoted to act as a wrapper for Product.

### Why are these changes introduced?

This PR came out of a conversation with @tyleralsbury earlier this year while working on [adding support for combined listings / 2k variants](https://github.com/Shopify/dawn/pull/3246). In that PR, the `VariantSelects` class needs to swap out the page content for its product when the selected variant is associated with a different Combined Listing child product. `VariantSelects` can be embedded in a few different product contexts (main product, featured product, quick add modal), and I was running into an issue where the class needed some pretty janky logic to determine its wrapping context ([here](https://github.com/Shopify/dawn/pull/3289/files#diff-47a8b2f1172e03da15783b379f0e007d8e3dfd27cbd9142adf9e32c4b9591079L1152-L1158) and [here](https://github.com/Shopify/dawn/pull/3289/files#diff-47a8b2f1172e03da15783b379f0e007d8e3dfd27cbd9142adf9e32c4b9591079L1170-L1172)). For a more stable architecture, child components shouldn't care what context they are rendered in.

An additional problem is that if components need to be updated when an event in a sibling happens, there wasn't a great "namespaced" context to manage the update. We had two primary patterns for handling this:

**Directly update sibling content** ([link](https://github.com/Shopify/dawn/blob/05e61a33d45358b09b85a089e0dd1b19109320c9/assets/global.js#L1092-L1102))
```js
class VariantSelects extends HTMLElement {
  
  onVariantChange(event) {
    ...
    this.updatePickupAvailability();
    ...
  }

  updatePickupAvailability() {
    const pickUpAvailability = document.querySelector('pickup-availability');
    if (!pickUpAvailability) return;

    if (this.currentVariant && this.currentVariant.available) {
      pickUpAvailability.fetchAvailability(this.currentVariant.id);
    } else {
      pickUpAvailability.removeAttribute('available');
      pickUpAvailability.innerHTML = '';
    }
  }
}
```

**Publish an event with the section, sibling subscription triggers update** ([link](https://github.com/Shopify/dawn/blob/05e61a33d45358b09b85a089e0dd1b19109320c9/assets/global.js#L1183-L1189))
```js
class VariantSelects extends HTMLElement {
    ...
    publish(PUB_SUB_EVENTS.variantChange, {
      data: {
        sectionId,
        html,
        variant: this.currentVariant,
      },
    });
  ...
}

class ProductInfo extends HTMLElement {
  connectedCallback() {
      this.variantChangeUnsubscriber = subscribe(PUB_SUB_EVENTS.variantChange, (event) => {
      const sectionId = this.dataset.originalSection ? this.dataset.originalSection : this.dataset.section;

      // check whether the event came from this ProductInfo's section
      if (event.data.sectionId !== sectionId) return;
      this.updateQuantityRules(event.data.sectionId, event.data.html);
      this.setQuantityBoundries();
    });
  }
}
```

### What approach did you take?

We have 2 primary issues: child components need to be context aware, and handling product component updates was complicated and inconsistent.

To solve both of these problems, I introduced a product class that wraps the entire rendered product and serves as an event handling puppeteer.

We already had a product wrapper class called `ProductInfo` that wrapped part of the rendered product; I moved it higher up the tree and extracted component update logic into it.

### Other considerations

I left a [comment](https://github.com/Shopify/dawn/pull/3289/files#r1499947293), but we possibly could decompose further by moving the old ProductInfo content into a QuantityForm class.

### Visual impact on existing themes

None

### Testing steps/scenarios

Here is a zip file for these theme changes: https://github.com/Shopify/dawn/archive/refs/heads/add-product-wrapper-class.zip

- [ ] Option value switching still works as expected
  - [ ] Switching between option values
  - [ ] Switching to an option value that has no associated variant
  - [ ] Switching between combined listings children
- [ ] Product recommendations load when switching between combined listings children
- [ ] Quick order list is updated when switching between combined listings children
- [ ] Quick add modal
  - [ ] works as expected for a normal product
  - [ ] works as expected for a combined listing product
- [ ] Quantity and volume price rules. (A catalog is set up on the shop, feel free to add yourself as a user or reach out to me for a test account)

### Demo links

- https://combined-listings-test.myshopify.com/products/chinos-orange

### Checklist
- [x] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [x] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [x] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [x] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [x] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [x] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

File: assets/global.js [hunk 1]
```diff
@@ -995,75 +995,35 @@ customElements.define('slideshow-component', SlideshowComponent);
 class VariantSelects extends HTMLElement {
   constructor() {
     super();
-    this.addEventListener('change', this.handleProductUpdate);
-    this.initializeProductSwapUtility();
   }
```
Comment (lhoffbeck) [inline assets/global.js:999]: This is the primary change of this PR--extracting product-update logic out of VariantSelects to a wrapper class. This also enables us to eliminate some gnarly selectors from [here](https://github.com/Shopify/dawn/pull/3289/files#diff-47a8b2f1172e03da15783b379f0e007d8e3dfd27cbd9142adf9e32c4b9591079L1152-L1158) and [here](https://github.com/Shopify/dawn/pull/3289/files#diff-47a8b2f1172e03da15783b379f0e007d8e3dfd27cbd9142adf9e32c4b9591079L1170-L1172).

File: assets/product-info.js [hunk 2]
```diff
@@ -2,81 +2,324 @@ if (!customElements.get('product-info')) {
...
+      onVariantChangeUnsubscriber = undefined;
+      cartUpdateUnsubscriber = undefined;
+      swapProductUtility = undefined;
+      abortController = undefined;
+
       constructor() {
         super();
-        this.input = this.querySelector('.quantity__input');
-        this.currentVariant = this.querySelector('.product-variant-id');
-        this.submitButton = this.querySelector('[type="submit"]');
```
Comment (lhoffbeck) [inline assets/product-info.js:9]: `currentVariant` can be grabbed from the product form class, `submitButton` wasn't used anywhere.

File: assets/quick-order-list.js
Comment (lhoffbeck) [inline assets/quick-order-list.js:1]: This content update now happens declaratively in ProductInfo

File: assets/global.js [hunk 3]
```diff
@@ -1315,39 +1085,24 @@ class ProductRecommendations extends HTMLElement {
...
 
   connectedCallback() {
```
Comment (lhoffbeck) [inline assets/global.js:1087]: Recommendation reinitialization now happens declaratively in the ProductInfo class

**Review by tauthomas01 — COMMENTED** [review]
There are a few things I noticed while tophatting:

1) When I use tab navigation, navigating through options with regular variant switching works well (e.g. `?variant=47651697099054`), but when I load new child product (e.g. `chinos-black?variant=xxx`), the focus is set after the main product section. Since it's loaded asynchronously, it would be nice to set the focus before the main product section so we're not losing out sight when a new content is loaded.

https://github.com/Shopify/dawn/assets/658169/fa56a1a9-c29c-4566-9557-7813f38b4c09

2) There is a JS error in some case where the cart section id looks undefined when adding the product into the cart. It says the following:

```
TypeError: Cannot read properties of null (reading 'innerHTML')
    at CartNotification.getSectionInnerHTML(cart-notification.js:66:85)
```

I looked into this and it's related to this:

```
getSectionInnerHTML(html, selector = '.shopify-section') {
    return new DOMParser().parseFromString(html, 'text/html').querySelector(selector).innerHTML;
  }
```

Printing the `selector` param returns this: `'[id="cart-notification-product-undefined"]'`

https://github.com/Shopify/dawn/assets/658169/0ad053a0-610c-4154-80c7-b415e5f6b50e

File: assets/product-form.js [hunk 4]
```diff
@@ -113,6 +114,20 @@ if (!customElements.get('product-form')) {
...
+
+      toggleSubmitButton(disable = true, text) {
+        if (disable) {
+          this.submitButton.setAttribute('disabled', 'disabled');
+          if (text) this.submitButtonText.textContent = text;
+        } else {
+          this.submitButton.removeAttribute('disabled');
+          this.submitButtonText.textContent = window.variantStrings.addToCart;
+        }
+      }
```
Comment (lhoffbeck) [inline assets/product-form.js:126]: `toggleSubmitButton` logic was pulled out of the VariantSelects class

File: sections/main-product.liquid [hunk 5]
```diff
@@ -1,7 +1,10 @@
-<section
+<product-info
   id="MainProduct-{{ section.id }}"
   class="section-{{ section.id }}-padding gradient color-{{ section.settings.color_scheme }}"
   data-section="{{ section.id }}"
+  data-product-id="{{ product.id }}"
+  data-update-url="true"
+  data-url="{{ product.url }}"
 >
```
Comment (lhoffbeck) [inline sections/main-product.liquid:8]: Worth flagging, this node was changed from a `section` to a `product-info`

File: sections/main-product.liquid [hunk 6]
```diff
@@ -746,7 +747,7 @@
       }
     </script>
   </div>
-</section>
+</product-info>
```
Comment (ludoboludo) [inline sections/main-product.liquid:750]: This is somehow breaking the padding applied and the content is flush with the header at the top: 


Seems like it's missing a `display: block;` that it was getting from the user agent stylesheet:

Comment (ludoboludo) [reply sections/main-product.liquid:750]: Which should also fix the color scheme issue not working atm:

**Review by ludoboludo — CHANGES_REQUESTED** [review]
